pkg/cloudprovider/overlay/cloudprovider.go

```diff
@@ -0,0 +1,197 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package overlay
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/mitchellh/hashstructure/v2"
+	"github.com/samber/lo"
+	lop "github.com/samber/lo/parallel"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/apis/v1alpha1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/options"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+)
+
+type decorator struct {
+	cloudprovider.CloudProvider
+	kubeClient client.Client
+}
+
+// Decorate returns a new `CloudProvider` instance that will delegate the GetInstanceTypes
+// calls to the argument, `cloudProvider`, and provide instance types with NodeOverlays applied to them. The
+func Decorate(cloudProvider cloudprovider.CloudProvider, kubeClient client.Client) cloudprovider.CloudProvider {
+	return &decorator{CloudProvider: cloudProvider, kubeClient: kubeClient}
+}
+
+func (d *decorator) GetInstanceTypes(ctx context.Context, nodePool *v1.NodePool) ([]*cloudprovider.InstanceType, error) {
+	its, err := pullInstanceTypes(ctx, d.CloudProvider, nodePool)
+	if err != nil {
+		return []*cloudprovider.InstanceType{}, err
+	}
+	if options.FromContext(ctx).FeatureGates.NodeOverlay {
+		// The additional requirements will be added to the instance type during scheduling simulation
+		// Since getting instance types is done on a NodePool level, these requirnements were always assumed
+		// to be allowed with these instance types.
+		addNodePoolRequirements(nodePool, its)
+		its, err = OverlayInstanceTypes(ctx, d.kubeClient, its)
+		if err != nil {
+			return []*cloudprovider.InstanceType{}, err
+		}
+		removeNodePoolRequirements(nodePool, its)
+	}
+	return its, nil
+}
+
+func OverlayInstanceTypes(ctx context.Context, kubeClient client.Client, instanceTypes []*cloudprovider.InstanceType) ([]*cloudprovider.InstanceType, error) {
+	overlays := &v1alpha1.NodeOverlayList{}
+	err := kubeClient.List(ctx, overlays)
+	if err != nil {
+		return []*cloudprovider.InstanceType{}, err
+	}
+
+	// Only apply overlays that have passed runtime validation and know there have not been
+	// any updates to the spec of the overlay field
+	validOverlays := lo.Filter(overlays.Items, func(o v1alpha1.NodeOverlay, _ int) bool {
+		condition := o.StatusConditions().Get(v1alpha1.ConditionTypeValidationSucceeded)
+		return condition.IsTrue() && condition.ObservedGeneration == o.Generation
```

Comment:
I brought this up in the validation PR as well, but I'm wondering if we need some concept of "the observed generation of the group of overlays". I'm thinking about the scenario where we only successfully update the status for a subset of overlays, so there may be overlays with conflicts that we're unaware of. Without that group observed generation construct, we wouldn't be able to treat the conditions as a source of truth here (at least if I've thought it through correctly).